### ISSUE TITLE
The wrong imp of the inner-product operation

### ISSUE THREAD

**Issue by XiaLiPKU** [leaderj1001/Stand-Alone-Self-Attention#10]
In Equation 2 of the paper, the query and the key are fed into inner-product operation, instead of point multiplication.

So the follow line
https://github.com/leaderj1001/Stand-Alone-Self-Attention/blob/e0a168ef8d4a7b93ae706a7d7c68b982e112821e/attention.py#L48
should be
`out = (q_out * k_out).sum(dim=2)`

**Comment by 20171130** [comment]
I found the same problem. It seems the implementation in the code is equivalent to having #attention heads = #embed dimensions.

**Comment by ifeherva** [comment]
@XiaLiPKU How would that modify line 49 and 50?

**Comment by canaltin** [comment]
@20171130 That was also my first opinion, but then there is an inconsistency with "groups" definition  (to replicate the "attention heads") throughout the paper & the code.  

Anyway, your alternative implementation helped me to understand the general concepts:
https://github.com/20171130/AttentionLite/blob/master/model.py

**Comment by KinWaiCheuk** [comment]
Shouldn't we implement the inner product using `torch.matmul`?
To me, the fix should look like this:
```python
    def forward(self, x):
        batch, channels, height, width = x.size()

        padded_x = F.pad(x, [self.padding, self.padding, self.padding, self.padding])
        q_out = self.query_conv(x)
        k_out = self.key_conv(padded_x)
        v_out = self.value_conv(padded_x)
        
        k_out = k_out.unfold(2, self.kernel_size, self.stride).unfold(3, self.kernel_size, self.stride)
        # (batch, channels, H, W, H_local_w, W_local_w) 
        
        v_out = v_out.unfold(2, self.kernel_size, self.stride).unfold(3, self.kernel_size, self.stride)
        # (batch, channels, H, W, H_local_w, W_local_w) 

        k_out_h, k_out_w = k_out.split(self.out_channels // 2, dim=1)
        
        k_out = torch.cat((k_out_h + self.rel_h, k_out_w + self.rel_w), dim=1) # relative position?

        k_out = k_out.contiguous().view(batch, self.groups, self.out_channels // self.groups, height, width, -1)
        v_out = v_out.contiguous().view(batch, self.groups, self.out_channels // self.groups, height, width, -1)
        # (batch, n_heads, feature_per_head, H, W, local H X W)
        
        k_out = k_out.permute(0,1,3,4,2,5) # rearranging dims for QK multiplication later
        # (batch, n_heads, H, W, feature_per_head, local H X W)
        
        v_out = v_out.permute(0,1,3,4,5,2) # rearranging dims for (att,v_out) later
        
        # expand the last dimension s.t. it can multiple with the local att window
        q_out = q_out.view(batch, self.groups, self.out_channels // self.groups, height, width, 1)
        # (batch, n_heads, feature_per_head, H, W, 1)
        q_out = q_out.permute(0,1,3,4,5,2) # rearranging dims for QK multiplication later
        # (batch, n_heads, H, W, 1, feature_per_head)
        
        energy = torch.matmul(q_out,k_out)     
        attention = F.softmax(energy, dim=-1)
        # (batch, n_heads, H, W, 1, local H X W)
        
        out = torch.matmul(attention, v_out)
        # (batch, n_heads, H, W, 1, feature_per_head)
        out = out.permute(0,1,5,2,3,4)
        # (batch, n_heads, feature_per_head, H, W, 1)
        out = out.reshape(batch, self.out_channels, height, width)
        # (batch, c, H, W)
        
        return out, attention.squeeze(4)
```
